core/src/main/java/org/pentaho/di/core/database/SnowflakeHVDatabaseMeta.java

```diff
@@ -0,0 +1,464 @@
+ * you may not use this file except in compliance with
+ * the License. You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *
+ ******************************************************************************/
+
+package org.pentaho.di.core.database;
+
+import com.google.common.base.Preconditions;
+import org.pentaho.di.core.Const;
+import org.pentaho.di.core.row.ValueMetaInterface;
+
+import java.sql.DatabaseMetaData;
+import java.sql.ResultSet;
+import java.sql.SQLException;
+import java.util.HashMap;
+import java.util.Map;
+
+import static org.pentaho.di.core.row.ValueMetaInterface.TYPE_BIGNUMBER;
+import static org.pentaho.di.core.row.ValueMetaInterface.TYPE_BINARY;
+import static org.pentaho.di.core.row.ValueMetaInterface.TYPE_BOOLEAN;
+import static org.pentaho.di.core.row.ValueMetaInterface.TYPE_DATE;
+import static org.pentaho.di.core.row.ValueMetaInterface.TYPE_INTEGER;
+import static org.pentaho.di.core.row.ValueMetaInterface.TYPE_NUMBER;
+import static org.pentaho.di.core.row.ValueMetaInterface.TYPE_STRING;
+import static org.pentaho.di.core.row.ValueMetaInterface.TYPE_TIMESTAMP;
+import static org.pentaho.di.core.util.Utils.isEmpty;
+
+@SuppressWarnings ( "unused" )
+public class SnowflakeHVDatabaseMeta extends BaseDatabaseMeta implements DatabaseInterface {
+
+  private static final String ALTER_TABLE = "ALTER TABLE ";
+  public static final String WAREHOUSE = "warehouse";
+
+  @Override
+  public int[] getAccessTypeList() {
+    return new int[] { DatabaseMeta.TYPE_ACCESS_NATIVE, DatabaseMeta.TYPE_ACCESS_JNDI };
+  }
+
+  @Override
+  public int getDefaultDatabasePort() {
+    if ( getAccessType() == DatabaseMeta.TYPE_ACCESS_NATIVE ) {
+      return 443;
+    }
+    return -1;
+  }
+
+  @Override
+  public Map<String, String> getDefaultOptions() {
+    Map<String, String> defaultOptions = new HashMap<>();
+    defaultOptions.put( getPluginId() + ".ssl", "on" );
+
+    return defaultOptions;
+  }
+
+  @Override
+  public boolean isFetchSizeSupported() {
+    return false;
+  }
+
+  @Override
+  public String getDriverClass() {
+    return "com.snowflake.client.jdbc.SnowflakeDriver";
+  }
+
+  @Override
+  public String getURL( String hostname, String port, String databaseName ) {
+    Preconditions.checkArgument( !isEmpty( hostname ) );
+    Preconditions.checkArgument( !isEmpty( databaseName ) );
+
+    String realHostname = hostname;
+    String account = hostname;
+    if ( !realHostname.contains( "." ) ) {
+      realHostname = hostname + ".snowflakecomputing.com";
+    } else {
+      account = hostname.substring( 0, hostname.indexOf( '.' ) );
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

If the indexOf() search on line 92 does not find the substring and returns -1, the call to substring() on line 92 will throw IndexOutOfBoundsException. Note that the method is publicly accessible to external callers who might not be aware of the constraint on string hostname.